File: sciurus17_examples/scripts/head_camera_tracking.py
#!/usr/bin/env python
# coding: utf-8

# Copyright 2019 RT Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import rospy
import math
import sys

# for ObjectTracker
import cv2
import numpy as np
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError
from geometry_msgs.msg import Point

# for NeckYawPitch
import actionlib
from control_msgs.msg import (
    FollowJointTrajectoryAction,
    FollowJointTrajectoryGoal,
    JointTrajectoryControllerState
)
from trajectory_msgs.msg import JointTrajectoryPoint

class ObjectTracker:
    def __init__(self):
        self._bridge = CvBridge()
        self._image_sub = rospy.Subscriber("/sciurus17/camera/color/image_raw", Image, self._image_callback, queue_size=1)
        self._image_pub = rospy.Publisher("~output_image", Image, queue_size=1)
        self._object_rect = [0,0,0,0]
        self._image_shape = Point()
        self._object_detected = False

        self._CV_MAJOR_VERSION, _, _ = cv2.__version__.split('.')

        # カスケードファイルの読み込み
        # 例
        # self._face_cascade = cv2.CascadeClassifier("/home/USER_NAME/.local/lib/python2.7/site-packages/cv2/data/haarcascade_frontalface_alt2.xml")
        # self._eyes_cascade = cv2.CascadeClassifier("/home/USER_NAME/.local/lib/python2.7/site-packages/cv2/data/haarcascade_eye.xml")
        self._face_cascade = ""
        self._eyes_cascade = ""


    def _image_callback(self, ros_image):
        try:
            input_image = self._bridge.imgmsg_to_cv2(ros_image, "bgr8")
        except CvBridgeError as e:
            rospy.logerr(e)
            
        # 画像のwidth, heightを取得
        self._image_shape.x = input_image.shape[1]
        self._image_shape.y = input_image.shape[0]

        # オブジェクト(特定色 or 顔) の検出
        output_image = self._detect_orange_object(input_image)
        # output_image = self._detect_blue_object(input_image)
        # output_image = self._detect_face(input_image)

        self._image_pub.publish(self._bridge.cv2_to_imgmsg(output_image, "bgr8"))


    def get_object_position(self):
        # 画像中心を0, 0とした座標系におけるオブジェクトの座標を出力
        # オブジェクトの座標は-1.0 ~ 1.0に正規化される

        object_center = Point(
                self._object_rect[0] + self._object_rect[2] * 0.5,
                self._object_rect[1] + self._object_rect[3] * 0.5,
                0)

        # 画像の中心を0, 0とした座標系に変換
        translated_point = Point()
        translated_point.x = object_center.x - self._image_shape.x * 0.5
        translated_point.y = -(object_center.y - self._image_shape.y * 0.5)

        # 正規化
        normalized_point = Point()
        if self._image_shape.x != 0 and self._image_shape.y != 0:
            normalized_point.x = translated_point.x / (self._image_shape.x * 0.5)
            normalized_point.y = translated_point.y / (self._image_shape.y * 0.5)

        return normalized_point


    def object_detected(self):
        return self._object_detected


    def _detect_color_object(self, bgr_image, lower_color, upper_color):
        # 画像から指定された色の物体を検出する

        MIN_OBJECT_SIZE = 1000 # px * px

        # BGR画像をHSV色空間に変換
        hsv = cv2.cvtColor(bgr_image, cv2.COLOR_BGR2HSV)

        # 色を抽出するマスクを生成
        mask = cv2.inRange(hsv, lower_color, upper_color)

        # マスクから輪郭を抽出
        if self._CV_MAJOR_VERSION == '4':
            contours, _ = cv2.findContours(mask, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
        else:
            _, contours, _ = cv2.findContours(mask, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)

        # 輪郭を長方形に変換し、配列に格納
        rects = []
        for contour in contours:
            approx = cv2.convexHull(contour)
            rect = cv2.boundingRect(approx)
            rects.append(rect)

        self._object_detected = False
        if len(rects) > 0:
            # 最も大きい長方形を抽出
            rect = max(rects, key=(lambda x: x[2] * x[3]))

            # 長方形が小さければ検出判定にしない
            if rect[2] * rect[3] > MIN_OBJECT_SIZE:
                # 抽出した長方形を画像に描画する
                cv2.rectangle(bgr_image, 
                        (rect[0], rect[1]), 
                        (rect[0] + rect[2], rect[1] + rect[3]), 
                        (0, 0, 255), thickness=2)

                self._object_rect = rect
                self._object_detected = True

        return bgr_image


    def _detect_orange_object(self, bgr_image):
        # H: 0 ~ 179 (0 ~ 360°)
        # S: 0 ~ 255 (0 ~ 100%)
        # V: 0 ~ 255 (0 ~ 100%)
        lower_orange = np.array([5,127,127])
        upper_orange = np.array([20,255,255])

        return self._detect_color_object(bgr_image, lower_orange, upper_orange)


    def _detect_blue_object(self, bgr_image):
        # H: 0 ~ 179 (0 ~ 360°)
        # S: 0 ~ 255 (0 ~ 100%)
        # V: 0 ~ 255 (0 ~ 100%)
        lower_blue = np.array([100,127,127])
        upper_blue = np.array([110,255,255])

        return self._detect_color_object(bgr_image, lower_blue, upper_blue)


    def _detect_face(self, bgr_image):
        # 画像から顔(正面)を検出する

        SCALE = 4

        # カスケードファイルがセットされているかを確認
        if self._face_cascade == "" or self._eyes_cascade == "":
            rospy.logerr("cascade file does not set")
            return bgr_image

        # BGR画像をグレー画像に変換
        gray = cv2.cvtColor(bgr_image, cv2.COLOR_BGR2GRAY)

        # 処理時間短縮のため画像を縮小
        height, width = gray.shape[:2]
        small_gray = cv2.resize(gray, (width/SCALE, height/SCALE))

        # カスケードファイルを使って顔認識
        small_faces = self._face_cascade.detectMultiScale(small_gray)

        self._object_detected = False
        for small_face in small_faces:
            # 顔の領域を元のサイズに戻す
            face = small_face*SCALE
            
            # グレー画像から顔部分を抽出
            roi_gray = gray[
                    face[1]:face[1]+face[3],
                    face[0]:face[0]+face[2]]

            # 顔の中から目を検知
            eyes = self._eyes_cascade.detectMultiScale(roi_gray)

            # 目を検出したら、対象のrect(座標と大きさ)を記録する
            if len(eyes) > 0:
                cv2.rectangle(bgr_image, 
                        (face[0],face[1]), 
                        (face[0]+face[2], face[1]+face[3]), 
                        (0,0,255),2)

                self._object_rect = face
                self._object_detected = True
                break

        return bgr_image


class NeckYawPitch(object):
    def __init__(self):
        self.__client = actionlib.SimpleActionClient("/sciurus17/controller3/neck_controller/follow_joint_trajectory",
                                                     FollowJointTrajectoryAction)
        self.__client.wait_for_server(rospy.Duration(5.0))
        if not self.__client.wait_for_server(rospy.Duration(5.0)):
            rospy.logerr("Action Server Not Found")
            rospy.signal_shutdown("Action Server not found")
            sys.exit(1)

        self._state_sub = rospy.Subscriber("/sciurus17/controller3/neck_controller/state", 
                JointTrajectoryControllerState, self._state_callback, queue_size=1)

        self._state_received = False
        self._current_yaw = 0.0 # Degree
        self._current_pitch = 0.0 # Degree


    def _state_callback(self, state):
        # 首の現在角度を取得

        self._state_received = True
        yaw_radian = state.actual.positions[0]
        pitch_radian = state.actual.positions[1]

        self._current_yaw = math.degrees(yaw_radian)
        self._current_pitch = math.degrees(pitch_radian)


    def state_received(self):
        return self._state_received


    def get_current_yaw(self):
        return self._current_yaw


    def get_current_pitch(self):
        return self._current_pitch


    def set_angle(self, yaw_angle, pitch_angle, goal_secs=1.0e-9):
        # 首を指定角度に動かす
        goal = FollowJointTrajectoryGoal()
        goal.trajectory.joint_names = ["neck_yaw_joint", "neck_pitch_joint"]

        yawpoint = JointTrajectoryPoint()
        yawpoint.positions.append(yaw_angle)
        yawpoint.positions.append(pitch_angle)
        yawpoint.time_from_start = rospy.Duration(goal_secs)
        goal.trajectory.points.append(yawpoint)

        self.__client.send_goal(goal)
        self.__client.wait_for_result(rospy.Duration(0.1))
        return self.__client.get_result()


def hook_shutdown():
    # shutdown時に0度へ戻る
    neck.set_angle(math.radians(0), math.radians(0), 3.0)


def main():
    r = rospy.Rate(60)

    rospy.on_shutdown(hook_shutdown)

    # オブジェクト追跡のしきい値
    # 正規化された座標系(px, px)
    THRESH_X = 0.05
    THRESH_Y = 0.05

    # 首の初期角度 Degree
    INITIAL_YAW_ANGLE = 0
    INITIAL_PITCH_ANGLE = 0

    # 首の制御角度リミット値 Degree
    MAX_YAW_ANGLE   = 120
    MIN_YAW_ANGLE   = -120
    MAX_PITCH_ANGLE = 50
    MIN_PITCH_ANGLE = -70

    # 首の制御量
    # 値が大きいほど首を大きく動かす
    OPERATION_GAIN_X = 5.0
    OPERATION_GAIN_Y = 5.0

    # 初期角度に戻る時の制御角度 Degree
    RESET_OPERATION_ANGLE = 3

    # 現在の首角度を取得する
    # ここで現在の首角度を取得することで、ゆっくり初期角度へ戻る
    while not neck.state_received():
        pass
    yaw_angle = neck.get_current_yaw()
    pitch_angle = neck.get_current_pitch()

    look_object = False
    detection_timestamp = rospy.Time.now()

    while not rospy.is_shutdown():
        # 正規化されたオブジェクトの座標を取得
        object_position = object_tracker.get_object_position()

        if object_tracker.object_detected():
            detection_timestamp = rospy.Time.now()
            look_object = True
        else:
            lost_time = rospy.Time.now() - detection_timestamp
            # 一定時間オブジェクトが見つからない場合は初期角度に戻る
            if lost_time.to_sec() > 1.0:
                look_object = False

        if look_object:
            # オブジェクトが画像中心にくるように首を動かす
            if math.fabs(object_position.x) > THRESH_X:
                yaw_angle += -object_position.x * OPERATION_GAIN_X

            if math.fabs(object_position.y) > THRESH_Y:
                pitch_angle += object_position.y * OPERATION_GAIN_Y

            # 首の制御角度を制限する
            if yaw_angle > MAX_YAW_ANGLE:
                yaw_angle = MAX_YAW_ANGLE
            if yaw_angle < MIN_YAW_ANGLE:
                yaw_angle = MIN_YAW_ANGLE

            if pitch_angle > MAX_PITCH_ANGLE:
                pitch_angle = MAX_PITCH_ANGLE
            if pitch_angle < MIN_PITCH_ANGLE:
                pitch_angle = MIN_PITCH_ANGLE

        else:
            # ゆっくり初期角度へ戻る
            diff_yaw_angle = yaw_angle - INITIAL_YAW_ANGLE
            if math.fabs(diff_yaw_angle) > RESET_OPERATION_ANGLE:
                yaw_angle -= math.copysign(RESET_OPERATION_ANGLE, diff_yaw_angle)
            else:
                yaw_angle = INITIAL_YAW_ANGLE

            diff_pitch_angle = pitch_angle - INITIAL_PITCH_ANGLE
            if math.fabs(diff_pitch_angle) > RESET_OPERATION_ANGLE:
                pitch_angle -= math.copysign(RESET_OPERATION_ANGLE, diff_pitch_angle)
            else:
                pitch_angle = INITIAL_PITCH_ANGLE

        neck.set_angle(math.radians(yaw_angle), math.radians(pitch_angle))

        r.sleep()


if __name__ == '__main__':
    rospy.init_node("head_camera_tracking")

    neck = NeckYawPitch()
    object_tracker = ObjectTracker()

    main()

File: sciurus17_examples/src/head_camera_tracking.cpp
// Copyright 2023 RT Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Reference:
// https://www.opencv-srf.com/2010/09/object-detection-using-color-seperation.html

#include <cmath>
#include <iostream>
#include <iomanip>
#include <memory>
#include "rclcpp_action/rclcpp_action.hpp"
#include "angles/angles.h"

#include "std_msgs/msg/float64_multi_array.hpp"
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/point_stamped.hpp"
#include "sensor_msgs/msg/camera_info.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "control_msgs/msg/joint_trajectory_controller_state.hpp"
#include "opencv2/opencv.hpp"
#include "cv_bridge/cv_bridge.h"
#include "control_msgs/action/follow_joint_trajectory.hpp"
using std::placeholders::_1;
using namespace std::chrono_literals;

class ColorDetection : public rclcpp::Node
{
public:

  ColorDetection()
  : Node("color_detection")
  {
    image_subscription_ = this->create_subscription<sensor_msgs::msg::Image>(
      "/head_camera/color/image_raw", 10, std::bind(&ColorDetection::image_callback, this, _1));

    image_thresholded_publisher_ =
      this->create_publisher<sensor_msgs::msg::Image>("image_thresholded", 10);

    object_point_publisher_ =
      this->create_publisher<geometry_msgs::msg::PointStamped>("object_detected", 10);
  }

private:
  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr image_subscription_;
  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr image_thresholded_publisher_;
  rclcpp::Publisher<geometry_msgs::msg::PointStamped>::SharedPtr object_point_publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
  cv::Point2d pixel_err_;
  bool has_object_point_ = false;
  // -1.0 ~ 1.0に正規化された検出位置
  cv::Point2d normalized_object_point_;

  void image_callback(const sensor_msgs::msg::Image::SharedPtr msg)
  {
    // オレンジ色の物体を検出するようにHSVの範囲を設定
    // 周囲の明るさ等の動作環境に合わせて調整
    const int LOW_H = 5, HIGH_H = 20;
    const int LOW_S = 120, HIGH_S = 255;
    const int LOW_V = 120, HIGH_V = 255;

    // ウェブカメラの画像を受け取る
    auto cv_img = cv_bridge::toCvShare(msg, msg->encoding);

    // 画像をRGBからHSVに変換
    cv::cvtColor(cv_img->image, cv_img->image, cv::COLOR_RGB2HSV);

    // 画像処理用の変数を用意
    cv::Mat img_thresholded;

    // 画像の二値化
    cv::inRange(
      cv_img->image,
      cv::Scalar(LOW_H, LOW_S, LOW_V),
      cv::Scalar(HIGH_H, HIGH_S, HIGH_V),
      img_thresholded);

    // ノイズ除去の処理
    cv::morphologyEx(
      img_thresholded,
      img_thresholded,
      cv::MORPH_OPEN,
      cv::getStructuringElement(cv::MORPH_RECT, cv::Size(5, 5)));

    // 穴埋めの処理
    cv::morphologyEx(
      img_thresholded,
      img_thresholded,
      cv::MORPH_CLOSE,
      cv::getStructuringElement(cv::MORPH_RECT, cv::Size(5, 5)));

    // 閾値による二値化画像を配信
    sensor_msgs::msg::Image::SharedPtr img_thresholded_msg =
      cv_bridge::CvImage(msg->header, "mono8", img_thresholded).toImageMsg();
    image_thresholded_publisher_->publish(*img_thresholded_msg);

    // 画像の検出領域におけるモーメントを計算
    cv::Moments moment = moments(img_thresholded);
    double d_m01 = moment.m01;
    double d_m10 = moment.m10;
    double d_area = moment.m00;

    // 検出領域のピクセル数が100000より大きい場合
    has_object_point_ = d_area > 100000;

    if (has_object_point_) {
      // 画像座標系における物体検出位置（2D）
      cv::Point2d object_point;
      object_point.x = d_m10 / d_area;
      object_point.y = d_m01 / d_area;

      RCLCPP_INFO_STREAM(this->get_logger(), "Detect at" << object_point << ".");

      // 画像の中心を原点とした検出位置に変換
      cv::Point2d translated_object_point;
      translated_object_point.x = object_point.x - msg->width / 2.0;
      translated_object_point.y = object_point.y - msg->height / 2.0;

      if (msg->width != 0 && msg->height) {
        normalized_object_point_.x = translated_object_point.x / msg->width / 2.0;
        normalized_object_point_.y = translated_object_point.y / msg->height / 2.0;
      }
      geometry_msgs::msg::PointStamped object_point_msg;
      object_point_msg.header = msg->header;
      object_point_msg.point.x = normalized_object_point_.x;
      object_point_msg.point.y = normalized_object_point_.y;
      object_point_publisher_->publish(object_point_msg);
    }
  }
};

class ObjectTracker : public rclcpp::Node
{
public:

  ObjectTracker()
  : Node("object_tracker")
  {
    timer_ = this->create_wall_timer(
      20ms, std::bind(&ObjectTracker::tracking, this));

    state_subscription_ = this->create_subscription<control_msgs::msg::JointTrajectoryControllerState>(
      "/neck_controller/controller_state", 10, std::bind(&ObjectTracker::state_callback, this, _1));

    object_point_subscription_ = this->create_subscription<geometry_msgs::msg::PointStamped>(
      "/object_detected", 10, std::bind(&ObjectTracker::point_callback, this, _1));

    angles_publisher_ =
      this->create_publisher<std_msgs::msg::Float64MultiArray>("/target_angles", 10);
  }

private:
  rclcpp::TimerBase::SharedPtr timer_;
  cv::Point2d pixel_err_;
  // -1.0 ~ 1.0に正規化された検出位置
  cv::Point2d normalized_object_point_;
  rclcpp::Subscription<control_msgs::msg::JointTrajectoryControllerState>::SharedPtr state_subscription_;
  rclcpp::Subscription<geometry_msgs::msg::PointStamped>::SharedPtr object_point_subscription_;
  rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr angles_publisher_;
  double current_yaw_angle_ = 0;
  double current_pitch_angle_ = 0;
  std::vector<double> current_angles_;
  geometry_msgs::msg::PointStamped::SharedPtr object_point_msg_;
  double yaw_angle_ = 0;
  double pitch_angle_ = 0;
  std::vector<double> target_angles_;

  void state_callback(const control_msgs::msg::JointTrajectoryControllerState::SharedPtr msg)
  {
    current_angles_ = msg->feedback.positions;
  }

  void point_callback(const geometry_msgs::msg::PointStamped::SharedPtr msg)
  {
    object_point_msg_ = msg;
  }

  void tracking()
  {
    // 追従を開始する物体位置の閾値
    const double THRESH_X = 0.05;
    const double THRESH_Y= 0.05;

    // 首角度の初期値
    const double INITIAL_YAW_ANGLE = 0;
    const double INITIAL_PITCH_ANGLE = 0;

    // 首可動範囲
    const double MAX_YAW_ANGLE = angles::from_degrees(120);
    const double MIN_YAW_ANGLE = angles::from_degrees(-120);
    const double MAX_PITCH_ANGLE = angles::from_degrees(50);
    const double MIN_PITCH_ANGLE = angles::from_degrees(-85);

    // 首角度初期化時の制御角度
    const double RESET_ANGLE_VEL = angles::from_degrees(0.3);

    // 物体が検出されなくなってから初期角度に戻り始めるまでの時間
    const std::chrono::nanoseconds DETECTION_TIMEOUT = 3s;

    // 首角度制御量
    // 値が大きいほど追従速度が速くなる
    const double OPERATION_GAIN_X = 0.05;
    const double OPERATION_GAIN_Y = 0.05;

    // 追従動作開始フラグ
    bool look_object = false;

    // 現在の首角度を取得
    if (current_angles_.empty()) {
      return;
    }
    if (target_angles_.empty()) {
      target_angles_ = current_angles_;
      yaw_angle_ = current_angles_[0];
      pitch_angle_ = current_angles_[1];
    }

    // 現在時刻
    auto now = this->get_clock()->now().nanoseconds();

    if (object_point_msg_) {
      // 物体を検出したとき追従動作開始フラグをtrueにする
      const auto detected_time = rclcpp::Time(object_point_msg_->header.stamp).nanoseconds();
      const auto POINT_ELAPSED_TIME = now - detected_time;
      look_object = POINT_ELAPSED_TIME < DETECTION_TIMEOUT.count();
    }

    // 物体が検出されたら追従を行う
    if (look_object) {
      // 物体検出位置を取得
      auto object_position = object_point_msg_->point;

      // 追従動作のための首角度を計算
      if (std::abs(object_position.x) > THRESH_X) {
        yaw_angle_ -= object_position.x * OPERATION_GAIN_X;
      }
      if (std::abs(object_position.y) > THRESH_Y) {
        pitch_angle_ -= object_position.y * OPERATION_GAIN_Y;
      }
    } else {
      // ゆっくりと初期角度へ戻る
      auto diff_yaw_angle = INITIAL_YAW_ANGLE - yaw_angle_;
      if (std::abs(diff_yaw_angle) > RESET_ANGLE_VEL) {
        yaw_angle_ += std::copysign(RESET_ANGLE_VEL, diff_yaw_angle);
      } else {
        yaw_angle_ = INITIAL_YAW_ANGLE;
      }

      auto diff_pitch_angle = INITIAL_PITCH_ANGLE - pitch_angle_;
      if (std::abs(diff_pitch_angle) > RESET_ANGLE_VEL) {
        pitch_angle_ += std::copysign(RESET_ANGLE_VEL, diff_pitch_angle);
      } else {
        pitch_angle_ = INITIAL_PITCH_ANGLE;
      }
    }

    // 目標首角度を制限角度内に収める
    yaw_angle_ = std::clamp(yaw_angle_, MIN_YAW_ANGLE, MAX_YAW_ANGLE);
    pitch_angle_ = std::clamp(pitch_angle_, MIN_PITCH_ANGLE, MAX_PITCH_ANGLE);

    // 目標角度に首を動かす
    std_msgs::msg::Float64MultiArray target_angles_msg;
    target_angles_msg.data.push_back(yaw_angle_);
    target_angles_msg.data.push_back(pitch_angle_);
    angles_publisher_->publish(target_angles_msg);
  }
};

class NeckControl : public rclcpp::Node
{
public:

  NeckControl()
  : Node("neck_control")
  {
    this->client_ptr_ = rclcpp_action::create_client<control_msgs::action::FollowJointTrajectory>(
      this,
      "neck_controller/follow_joint_trajectory");

    angles_subscription_ = this->create_subscription<std_msgs::msg::Float64MultiArray>(
      "/target_angles", 10, std::bind(&NeckControl::angles_callback, this, _1));
  }

private:
  rclcpp_action::Client<control_msgs::action::FollowJointTrajectory>::SharedPtr client_ptr_;
  rclcpp::Subscription<std_msgs::msg::Float64MultiArray>::SharedPtr angles_subscription_;
  control_msgs::msg::JointTrajectoryControllerState::SharedPtr neck_state_;

  void angles_callback(const std_msgs::msg::Float64MultiArray::SharedPtr msg) {
    using namespace std::placeholders;
    auto yaw_angle = msg->data[0];
    auto pitch_angle = msg->data[1];
    const double TIME_FROM_START = 0.001;

    if(!this->client_ptr_->wait_for_action_server(5s)) {
      RCLCPP_ERROR(this->get_logger(), "Action server not available after waiting");
      rclcpp::shutdown();
    }

    auto goal_msg = control_msgs::action::FollowJointTrajectory::Goal();
    goal_msg.trajectory.joint_names.push_back("neck_yaw_joint");
    goal_msg.trajectory.joint_names.push_back("neck_pitch_joint");

    trajectory_msgs::msg::JointTrajectoryPoint trajectory_point_msg;
    trajectory_point_msg.positions.push_back(yaw_angle);
    trajectory_point_msg.positions.push_back(pitch_angle);
    trajectory_point_msg.time_from_start = rclcpp::Duration::from_seconds(TIME_FROM_START);
    goal_msg.trajectory.points.push_back(trajectory_point_msg);

    this->client_ptr_->async_send_goal(goal_msg);
  }
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::NodeOptions node_options;
  rclcpp::WallRate loop_rate(20ms);
  node_options.automatically_declare_parameters_from_overrides(true);

  rclcpp::executors::MultiThreadedExecutor exec;
  auto color_detection_node = std::make_shared<ColorDetection>();
  auto neck_control_node = std::make_shared<NeckControl>();
  auto objeckt_tracker_node = std::make_shared<ObjectTracker>();
  exec.add_node(color_detection_node);
  exec.add_node(neck_control_node);
  exec.add_node(objeckt_tracker_node);
  exec.spin();

  rclcpp::shutdown();
  return 0;
}
